Joey Chen iP

JoeyChenSmart opened the pull request:
# my amazing project 💯💯🔥🔥
code from the heavens

haha jk they won't use java in heaven

Commit:
Add Gradle support

JoeyChenSmart committed:
Implement skeletal version that only echoes output

Changed the logo, and added a few functions to format system output
to stdout. Main function now also reads in and echoes input, and
exits when "bye" is entered. It will also print a goodbye message.

JoeyChenSmart committed:
Feat: Implement add, list features (Level-2)

firstly, created a constructor for Duke to make it easier to hold
state/variables. Also made some of the utility functions non-static.

added the `add` and `list` features. Uses an arraylist to store and
display strings input by the user. Created the functions `addItem`
and `printList` to carry this out.

minor fix: the indent function was not working properly for strings
with multiple lines.

note: will start using conventional commits style messages for
future commits.

JoeyChenSmart committed:
Feat: Implement tasks+mark as done (Level-3)

Added a new class `task` to keep track of tasks. Also replaced the
ArrayList<String> items with an ArrayList<Task>. Task can be pretty
printed with toString.

Added mark as done feature: Duke will look for strings beginning
with `done ` and parse the rest of the string into an int idx.
The task with index idx will be marked as done.
TODO: add error handling for strings with wrong formatting.

JoeyChenSmart committed:
Refactor(Task): Refactor into Task + Todo classes

Made Task an abstract class to facilitate future additions.
Each Task subclass now has an identifier which is prepended to
toString. The Todo class is a concrete implementation which
behaves similarly to the old Task class.

JoeyChenSmart committed:
Feat(Task): Add Deadline and Event classes

Added deadline and event classes which extend task.
Both classes have a new string parameter `time` to be set in the
constructor. This `time` string will be pretty printed in toString.

JoeyChenSmart committed:
Feat: Implement todo, event, deadline features (Level-4)

Added the above commands, parsing is done using very simple string
operations, but should suffice for our purposes.

Factored out the function `parseAndRun` from the `run` function in
Duke. This function should be further refactored in future updates.

Fixes: fixed minor string formatting issues in the event/deadline
classes. Also made the `addItem` message more verbose.

JoeyChenSmart committed:
Test: Add automated text UI testing

Added input+expected files, and modified run test script to compile
all .java files in the source folder.

Todo: use a proper snapshot test utility. Manually testing like this
is painful.

JoeyChenSmart committed:
Feat: added DukeException class

Added the custom exception class, with the nested enum of error
codes. Enum is great for this use case.

Todo: actually use this exception in Duke.java

JoeyChenSmart committed:
Feat: Add error handling (Level-5)

Duke now uses DukeExceptions to handle errors. Each exception
will have a default message. In Duke.run, we catch these
exceptions, print the message, and resume normal functions.

Also updated the snapshot tests since the error messages changed.

TODO: the code is getting very unwieldy. Really need to refactor
soon.

JoeyChenSmart committed:
Feat: Add delete feature (Level-6)

Added a delete feature. Implemenation is trivial since we are
already using an ArrayList, so a new function was added to
invoke the list's remove method.

JoeyChenSmart committed:
Test: Update text ui tests to last commit

UI tests did not include the delete feature. Added a few lines
to test for this.

JoeyChenSmart committed:
Refactor: Big refactoring of project

1. Created a TaskList class to manage task list related operations.
Moved the `markItem` and `deleteItem` methods from Duke there.

2. Created the Command enum and CommandExecutable functional
interface. each command will have a `match` and `dispatch` method
to respectively handle matching and executing the command.
There are many other nontrivial details that will be elaborated
on in the docs (coming soon).

3. Created the parser class to handle parsing/matching of the
input strings from the scanner and the commands.

4. Moved small misc functions out to Util class. These functions deal
mainly with reusable printing operations. Will change this to a full
blown UI class in the future.

JoeyChenSmart committed:
Refactor(Command): Use optionals in matcher

Previously matcher returned `null` for failed matches as a
pseudo-maybe (haskell maybe) value. Replaced this with an
actual optional value.

JoeyChenSmart committed:
Fix(Task): Remove unicode ticks and crosses

Task was using unicode characters for printing that misbehaved
sometimes (especially if the locale was changed). Replaced this with
the classic `[ ]` for 'unchecked' and `[X]` for 'checked' statuses.

JoeyChenSmart committed:
Merge pull request #1 from JoeyChenSmart/add-gradle-support

Add Gradle support

JoeyChenSmart committed:
Feat: Create Storage class

The new Storage class will use the Gson library to serialize/
deserialize the TaskList to and from a String respectively.

Created custom (de)serializers to deal with the polymorphic Task
objects.
Added a new public method `getItemsList` to TaskList to allow us
to insert new tasks directly without using the setter commands
(which print to stdout).

TODO: Find a 'smarter' way to create the TaskList object without
exposing the private ArrayList. Possibly add a new overloaded
constructor (which takes in a List).

TODO: Complete the functionality (save/load from a file), and
write some tests.

JoeyChenSmart committed:
Feat: Add Save and Load Features

Completed the save/load from file features.

Added 2 new  DukeExceptions for file io exceptions.

Added a new constructor to TaskList so we don't have to mainpulate
the internal ArrayList directly.

TODO: Add tests.

TODO: Creating a new DukeException requires writing a bit too much
boilerplate. Will change this in a future commit.

JoeyChenSmart committed:
Chore: Change save file name

Saved as duke.json instead of test.json. Saving as test.json
does not seem very professional.

JoeyChenSmart committed:
Feat: Add support for dates

Added basic support for dates using java's LocalDate class.
Also refactored out TimeTask from deadline and event.

JoeyChenSmart committed:
Chore: Fix gradle class path

Changed gradle's mainClassName variable to point to the correct
class.

JoeyChenSmart committed:
Merge pull request #2 from JoeyChenSmart/branch-Level-7

Branch level 7

JoeyChenSmart committed:
Merge branch 'master' into branch-Level-8

JoeyChenSmart committed:
Feat: Add date support for Storage class

Modified Storage class to work with the new dates.

JoeyChenSmart committed:
Merge pull request #3 from JoeyChenSmart/branch-Level-8

branch level 8

JoeyChenSmart committed:
Refactor(DukeException): Reduce boilerplate

Previously a new DukeException was created like this:
`throw new DukeException(DukeException.ERRORCODE)` which much
longer than it should be.

Added a new static convenience function to DukeException.Error
`create` that returns a new DukeException from that error code.

Also updated code to use this new function (instead of the old
way above).

Usage: `throw DukeException.ERRORCODE.create()`

JoeyChenSmart committed:
Refactor: Create Ui class + move command lambdas

1. Create Ui class to handle system i/o related function (printing,
scanner, etc). This Ui class replaces all the Util.* functions that
were being used. Ui is now passed as an argument in
CommandExecutable. Updated commands to use this argument.

2. Move command lambdas. The command lambdas (todoCommand,
listCommand etc) were strewn about in their respective classes
(Todo, Tasklist etc). Consolidated all of these commands as static
constants in a new CommandLibrary class. This will also deal with
any potential circular references.

JoeyChenSmart committed:
Refactor: Clean up code

Removed now unused Util class and resolved warnings in code (using
intelliJ).

JoeyChenSmart committed:
Refactor: Organize into packages

Moved everything into their respective package.

JoeyChenSmart committed:
Test: Add several JUnit tests

Added testing of task classes: Todo, Event and Deadline. Tests
check if the instance is created correctly, has the correct
identifier, prints correctly, and markAsComplete works.

Added tests for Command. Added 2 helper functions to help test
whether commands: reject bad input, and match inputs correctly.
Currently only tests the TODO command. Will add more tests if I
find time.

JoeyChenSmart committed:
Chore: Add Checkstyle

Added checkstyle to gradle and fixed some minor issues.

JoeyChenSmart committed:
Docs: Add JavaDocs to all public functions.

Wrote brief JavaDocs for all public functions. There were several
functions that should not have been private, and their access control
was modified according to their needs.

JoeyChenSmart committed:
Chore: Add extra checkstyle rules

Added rule to check that boolean variables/methods sound like
booleans.

Added rule to check for package declaration on first line of file.

Added rule to check that every public class has a JavaDoc header.

Will be adding more rules in the future (may even write custom
ones).

JoeyChenSmart committed:
Chore: Change access modifiers for some classes

Some classes are only accessed by their package and have been set
to package-private.

JoeyChenSmart committed:
docs: Add JavaDocs to public classes

Added brief descriptions for each public class.

JoeyChenSmart committed:
Feat: Add Find command (Level-9)

Added find command to filter and print TaskList. Made a getter in
Task to enable searching. Refactored TaskList's toString function
into a static enumerateItems function that is reused for the find
command.

JoeyChenSmart committed:
Merge pull request #4 from JoeyChenSmart/branch-A-JavaDoc

Merge JavaDoc branch

JoeyChenSmart committed:
Merge pull request #5 from JoeyChenSmart/branch-A-CodingStandard

Merge CodingStandard branch

JoeyChenSmart committed:
Merge branch 'master' into branch-Level-9

JoeyChenSmart committed:
Chore: Minor CheckStyle tweaks

Reformatted 1 line of imports.

CheckStyle's MissingJavadocType (looks for missing JavaDoc headers
for classes) is also flagging JUnit Test classes. Changed the
severity for this check to Warning since there's no other way
around this.

JoeyChenSmart committed:
Merge pull request #6 from JoeyChenSmart/branch-Level-9

Implement Find command

JoeyChenSmart committed:
Chore: Add more checkstyle rules

Added several more rules to make it easier to comply with the coding
guideline.

JoeyChenSmart committed:
Refactor(CommandLibrary): Change access modifiers

Command lambdas are only accessed within the Command package, so
theres no need for them to be public. Changed them to package
private.

JoeyChenSmart committed:
Refactor: Factored out Cli class from Ui

To facilitate creation of the new GUI, factored out Cli as an
implementation of the Ui interface, which the new GUI component
should implement.

JoeyChenSmart committed:
Feat: Add GUI

First iteration of working GUI. The new class GuiHelper is a
container for messages to and from duke. Pretty much a workaround
to let us reuse the existing code by implementing the Ui interface.

TODO: fix the problem where input/output text gets cut off, and
maybe add some styling

JoeyChenSmart committed:
Fix: Wrap long messages in GUI

Previously, long messages in the GUI were cut off. Added a function
to split the message up into chunks that can be displayed properly
one at a time.

JoeyChenSmart committed:
Docs: Add JavaDocs to new GUI classes

Also added to some other classes with missing JavaDocs.

JoeyChenSmart committed:
Merge pull request #7 from JoeyChenSmart/branch-Level-10

Add GUI support

JoeyChenSmart committed:
Fix: Text in GUI displaying incorrectly

Changed some properties in the fxml files to allow the text Labels to
automatically resize when the text is overrun. We don't need to split
up the messages any more.

JoeyChenSmart committed:
Docs: Add JavaDocs

Some JavaDocs were in a TODO state. Filled in docs for those.

Also removed an unused import.

JoeyChenSmart committed:
Style: Fixed several styling issues

Fixed styling issues according to CheckStyle. Mostly changing
constant names to UPPER_SNAKE_CASE and adding @Override annotations.

JoeyChenSmart committed:
Feat: Change Duke picture

Screenshotted the ASCII face for the CLI and saved it as an image for
Duke. Also increased the size of the image displayed.

afroneth commented on src/test/java/duke/command/CommandTest.java:
```diff
@@ -0,0 +1,38 @@
+import org.junit.jupiter.api.Test;
+
+public class CommandTest {
+    private void testCommandRejectsBadInputs(Command command) {
```
Perhaps you can consider using the following naming format "whatIsBeingTested_descriptionOfTestInputs_expectedOutcome" for your test cases?

JoeyChenSmart replied:
Good suggestion! Forgot about this style of naming tests.

afroneth commented on src/test/java/duke/task/TodoTest.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.junit.jupiter.api.Test;
+
+public class TodoTest {
```
Perhaps you could include more comprehensive test cases?

afroneth commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,13 @@
+package duke.ui;
```
Great idea to abstract the Ui class.

afroneth commented on src/main/java/duke/ui/GuiHelper.java:
```diff
@@ -0,0 +1,71 @@
+     * and output from the Command is not empty, Optional.empty() otherwise.
+     */
+    public Optional<List<String>> consumeCommandOutput() {
+        if (this.isNotConsumed) {
```
Clear implementation.

afroneth commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+     * @param idx Index of Task to markAsDone.
+     * @return Task of item that was just marked.
+     */
+    public Task markItem(int idx) {
```
Is it better to simply name the variable 'index' instead of 'idx'?

afroneth commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,138 @@
+            throw DukeException.Errors.FILE_READ_ERROR.create();
+        }
+    }
+
```
Impressive implementation! Although there are many lines of code, it is still very readable.

afroneth commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,56 @@
+                continue;
+            }
+            Matcher matcher = maybeMatcher.get();
+            // Should we throw a more specific error here?
```
Perhaps you should resolve this issue before you finalize your product?

JoeyChenSmart replied:
Good catch, I'll resolve this if I find time. Thanks for the review!

afroneth commented on src/main/java/duke/command/CommandLibrary.java:
```diff
@@ -0,0 +1,69 @@
+        ui.systemMessage(taskList.add(new Todo(description)));
+    };
+    static final CommandExecutable DELETE_COMMAND = (taskList, ui, args) -> {
+        int idx = Integer.parseInt(args[0]);
```
Perhaps naming the variable as 'index' is clearer?

JoeyChenSmart replied:
Thanks for pointing this out, didn't notice that I was using shorthands.

afroneth reviewed (commented):
Great job, the code is of high quality. Perhaps you can consider renaming your test case methods.

soaza commented on src/main/java/duke/ui/fxcomponents/MainWindow.java:
```diff
@@ -0,0 +1,88 @@
+        String input = userInput.getText();
+        this.guiHelper.setUserInput(input);
+        this.userMessage(input);
+        this.duke.nextIteration();
```
Might want to leave an empty line here for better clarity.

JoeyChenSmart replied:
Thanks for the review! Glad to know my code helped you. :)

Do you mind elaborating on which classes/methods you think needs more comments/javadocs?

soaza reviewed (commented):
Very well implemented design. Great use of lambdas,enums and abstraction. I learnt many things from reading your code, thanks! I would prefer if you add some comments for each class and methods(javadocs maybe?) for additional clarification. 
No violation of coding standards found by me. Overall, terrific job! LGTM.

JoeyChenSmart committed:
Feat: Add assertions

Add several assertions in the code. Note that most of the code does
not have many assumptions (apart from variables being non-null), and
the only place I've found they're relevant is in the Ui classes.

More non-null assertions may be useful when the variable passed is
not being used immediately, e.g. in constructors.

JoeyChenSmart committed:
Feat: Add more non-null assertions

Added some non-null assertions where they may be useful, e.g. in
constructors. Also removed some assertions where they won't, i.e.
in functions where an error will be thrown immediately if the
variable passed is null.

JoeyChenSmart committed:
Refactor: Minor renamings

Renamed several variables and methods according to suggestions from
PR review.

JoeyChenSmart committed:
Refactor(Storage): Refactor deserializer

Deserializer was violating SLAP. Factored out a helper method.

JoeyChenSmart committed:
Merge pull request #8 from JoeyChenSmart/branch-A-Assertions

Merge A-Assertions

JoeyChenSmart committed:
Merge pull request #9 from JoeyChenSmart/branch-A-CodeQuality

Merge A-CodeQuality

JoeyChenSmart committed:
Feat: Start using apache commons-cli

Changed the code to use apache commons-cli. Rather not reinvent the
wheel.

JoeyChenSmart committed:
Feat: Improve error messages

Use error/help messages generated by commons-cli.

JoeyChenSmart committed:
Refactor: Add Context class

Added a new context class that contains the TaskList and Ui from
duke. Easier to pass this around. This class is pretty much a record
type from sane languages.

JoeyChenSmart committed:
Feat: Add macro function

Added basic functionality for macros (barely working). Many important
things are still missing (see todos below).

TODO: add JavaDocs

TODO: clean up code using checkstyle

TODO: handle exceptions dealing with new macro creation

JoeyChenSmart committed:
Feat: Add exceptions to macro classes

Now throws new DukeExceptions when a macro definition has the
wrong format, and when macro execution encounters an error halfway
through execution.

JoeyChenSmart committed:
Fix: Fix error reporting for macros

Renamed DukeParseException to DukeCustomException to allow it to
be used in other places.

Errors that happen halfway in a macro were not being displayed
properly. This is due to the DukeCustomException's extraMessage
being overwritten. Fixed this by only allowing extraMessage to be
assigned when it is null.

JoeyChenSmart committed:
Style: Clean up issues using CheckStyle.

Cleaned up formatting issues and added copious JavaDocs.

JoeyChenSmart committed:
Merge pull request #10 from JoeyChenSmart/branch-C-FriendlierSyntax

Add macro support

JoeyChenSmart committed:
Docs(Parser): Add docs to getHelp

Forgot to add this in last push.

JoeyChenSmart committed:
Feat: Add default macros

According to the demands of the extension, there should also be
single letter shorthands for the commands. Added default macros to
fulfill this.

JoeyChenSmart committed:
Feat: Improve macro error messages.

Stopped the error message from printing "The following commands were
executed successfully..." when no commands were successful. Did the
same for the unsuccessful commands.

JoeyChenSmart committed:
Refactor(Macro): Refactor newMacro method

Applied more SLAP and extracted a method. Don't wanna risk losing
marks. Also added a comment for justification where I didn't apply
SLAP on principle (will make code unnecessarily long/hard to read).

JoeyChenSmart committed:
Docs: Clarify some javadocs

Made some javadocs/comments more understandable

JoeyChenSmart committed:
Docs: Credit reused code

Added comments to credit the module's tutorial code whenever i used
it. Rest of the code is Joey Chen OC which I am quite proud of.

JoeyChenSmart committed:
Fix: Fix unknown command error not appearing.

Forgot to add that part to the new parser. Also added some code to
ignore empty input from the user.

JoeyChenSmart committed:
Feat: Add window name

JoeyChenSmart committed:
Docs: Add user guide and product picture

Add a draft of the user guide. Will add more details later.

JoeyChenSmart committed:
Chore: Modify build.gradle

Removed the quality plugin that was being used as an experiment.

JoeyChenSmart committed:
Docs: Update user guide

Added more to the Macros section

JoeyChenSmart committed:
Chore: Rename Ui.png

forgot it was case sensitive

JoeyChenSmart committed:
Fix: Fix long text being clipped

reduced size of outputContainer so text is wrapped earlier.

JoeyChenSmart committed:
Update README.md
